### PR TITLE
#557 Josephine integration

### PR DESCRIPTION
**Pull request by radoslavzeman**
Separate table component
Create Next API to securely fetch Josephine data
Fetch Josephine data
Sort procurements data to show only new ones
Create pagination for the table
Snímka obrazovky (32)
Snímka obrazovky (33)
Snímka obrazovky (34)

### CODE REVIEW

File: next/.env.example [hunk 1]
```diff
@@ -12,6 +12,9 @@ NEXT_PUBLIC_MAPBOX_STYLE_ID=cm0l0wpy9007j01pj7smo6y6m
 NEXT_PUBLIC_MEILISEARCH_SEARCH_API_KEY=TODO
 NEXT_PUBLIC_MEILISEARCH_HOST=http://127.0.0.1:7700
 
+NEXT_PUBLIC_JOSEPHINE_API_TOKEN=TODO
```
Comment (radoslavzeman) [inline next/.env.example:15]: Should it really be `NEXT_PUBLIC_`?

File: strapi/src/components/sections/table.json
Comment (radoslavzeman) [inline strapi/src/components/sections/table.json:1]: This component should not be here, it was already removed.

File: strapi/src/components/sections/procurements.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    "description": ""
+  },
+  "options": {},
+  "attributes": {
+    "title": {
+      "type": "string"
+    },
+    "content": {
+      "type": "richtext"
+    },
```
Comment (radoslavzeman) [inline strapi/src/components/sections/procurements.json:14]: Please change to `text` of type string as it is in other sections.

File: strapi/src/api/page/content-types/page/schema.json
Comment (radoslavzeman) [inline strapi/src/api/page/content-types/page/schema.json:1]: Is this change in order of sections intended and discussed?

File: strapi/src/components/sections/procurements.json [hunk 3]
```diff
@@ -0,0 +1,19 @@
+{
+  "collectionName": "components_sections_procurements",
+  "info": {
+    "displayName": "Verejné obstarávanie",
```
Comment (radoslavzeman) [inline strapi/src/components/sections/procurements.json:4]: Please use plural: Verejné obstarávani**a**
e -> a

File: next/src/utils/formatPrice.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+/**
+ * Returns price in fixed format with two decimal places. Uses decimal point/comma depending on locale.
+ * Note: We don't use `style: 'currency' to keep € symbol at the end for simplicity.
+ * Example: 2.05 €
+ * Inspired by PAAS-MPA app
```
Comment (radoslavzeman) [inline next/src/utils/formatPrice.ts:5]: We usually add also a link to file on github. :)

File: next/src/services/josephine/utils.tsx [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        : ''
+
+    // format date
+    const formattedTenderFrom = new Date(tender.tender_rounds_mapped).toLocaleDateString('sk-SK', {
+      year: 'numeric',
+      month: 'numeric',
+      day: 'numeric',
+      hour: 'numeric',
+      minute: 'numeric',
+    })
```
Comment (radoslavzeman) [inline next/src/services/josephine/utils.tsx:68]: Please create a function for this in utils folder, similarly to other formatting functions.

File: next/src/services/josephine/fetchProcurements.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+ *
+ */
+
+export const fetchProcurements = async ({
+  timeframe,
+  currentPage = 1,
+  tendersPerPage = DEFAULT_TENDERS_PER_PAGE,
+}: FetchProcurementsProps) => {
+  const postFix = timeframe && timeframe !== 'all' ? `/${timeframe}` : ''
+  const fetchUrl = `${process.env.NEXT_JOSEPHINE_URL}${process.env.NEXT_JOSEPHINE_API_TOKEN}${postFix}`
```
Comment (radoslavzeman) [inline next/src/services/josephine/fetchProcurements.ts:57]: Please use less confusing name for envs - using `NEXT_` seems like useing `NEXT_PUBLIC_` prefix that changes env visibility for client.

If envs are not `NEXT_PUBLIC_` prefixed, we usually would use just `JOSEPHINE_URL` and `JOSEPHINE_API_TOKEN` as names.

File: next/src/services/josephine/fetchProcurements.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    ? tenderRounds?.sort(
+        (tenderA, tenderB) =>
+          new Date(tenderA.tender_round_to).getTime() - new Date(tenderB.tender_round_to).getTime(),
+      )[0]?.tender_round_to
+    : tenderRounds?.tender_round_to
+
+/**
+ * Fetches and parses client data from the Josephine client.
+ * It's wrapped into function to use it both directly and in API handler.
+ * @param timeframe - ['running', 'ended', 'all'] used for different fetch calls to Josephine API
```
Comment (radoslavzeman) [inline next/src/services/josephine/fetchProcurements.ts:43]: It would be nice to have an empty line before params for better readability. But that's maybe just my preference :)

File: next/src/pages/api/josephine.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  const { query } = _req
+
+  const options = {
+    timeframe:
+      isDefined(query?.timeframe) && typeof (query?.timeframe === 'string')
+        ? query?.timeframe.toString()
+        : '',
+    currentPage:
+      isDefined(query?.currentPage) && !Number.isNaN(Number(query?.currentPage))
+        ? Number(query?.currentPage)
```
Comment (radoslavzeman) [inline next/src/pages/api/josephine.ts:18]: same

File: next/src/pages/api/josephine.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        : 1,
+    tendersPerPage:
+      isDefined(query?.tendersPerPage) && !Number.isNaN(Number(query?.tendersPerPage))
+        ? Number(query?.tendersPerPage)
```
Comment (radoslavzeman) [inline next/src/pages/api/josephine.ts:22]: same

File: next/src/pages/api/josephine.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+export const DEFAULT_TENDERS_PER_PAGE = 5
+
+const handler = async (_req: NextApiRequest, res: NextApiResponse) => {
+  const { query } = _req
+
+  const options = {
+    timeframe:
+      isDefined(query?.timeframe) && typeof (query?.timeframe === 'string')
+        ? query?.timeframe.toString()
```
Comment (radoslavzeman) [inline next/src/pages/api/josephine.ts:14]: `?` is probably needed only in first `query?` usage

File: next/src/components/sections/ProcurementsSection.tsx [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+            rows={getRows(procurementsRunning, locale, t('procurements.detail'))}
+            visibleColumns={visibleColumns}
+            allColumns={allColumns}
+            headerAllColumns={headerAllColumns}
+          />
+        </div>
+      )}
+      <div className="flex justify-end">
+        <PaginationWithInput
+          currentPage={Number(currentPage)}
```
Comment (radoslavzeman) [inline next/src/components/sections/ProcurementsSection.tsx:79]: `currentPage` is already a number

File: next/src/components/sections/ProcurementsSection.tsx [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+          totalCount={
+            tendersPerPage && procurementsRunning?.tendersCount
```
Comment (radoslavzeman) [inline next/src/components/sections/ProcurementsSection.tsx:81]: Both conditions should be checked sooner.

Usually like this:
```tsx
{procurementsRunning?.tendersCount ? 
  <div className="flex justify-end">
        <PaginationWithInput
    ....
: null}
```

And for `tendersPerPage`, either define a default value already in object destructuring, or make it required in Strapi or something similar.

File: next/src/components/sections/ProcurementsSection.tsx [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+  const [currentPage, setCurrentPage] = useState(1)
+
+  const {
+    isPending: pendingRunning,
+    isError: isErrorRunning,
+    error: errorRunning,
+    data: procurementsRunning,
+  } = useQuery({
+    queryKey: ['Procurements running', tendersPerPage, currentPage], // TODO: add query key for pagination ['Procurement', currentPage]
```
Comment (radoslavzeman) [inline next/src/components/sections/ProcurementsSection.tsx:36]: We usually use `CamelCase` style for query keys

File: next/src/components/common/Table/Table.tsx
Comment (radoslavzeman) [inline next/src/components/common/Table/Table.tsx:1]: Didn't check this

**Comment by radoslavzeman** [comment]
Is `tender` and `procurement` the same thing in our case? Could we use consistent naming please?

File: next/.env.example (hunk 1 above)
Comment (jankaBehal) [reply next/.env.example:15]: you are right it should be without PUBLIC

File: next/src/components/sections/ProcurementsSection.tsx (hunk 13 above)
Comment (jankaBehal) [reply next/src/components/sections/ProcurementsSection.tsx:36]: done

File: next/src/services/josephine/fetchProcurements.ts (hunk 7 above)
Comment (jankaBehal) [reply next/src/services/josephine/fetchProcurements.ts:43]: according to JS Doc documentation it not specified, but we can have empty line if its better readable

File: next/src/services/josephine/utils.tsx (hunk 5 above)
Comment (jankaBehal) [reply next/src/services/josephine/utils.tsx:68]: I used one of the already created functions

File: next/src/utils/formatPrice.ts (hunk 4 above)
Comment (jankaBehal) [reply next/src/utils/formatPrice.ts:5]: added

File: strapi/src/api/page/content-types/page/schema.json
Comment (jankaBehal) [reply strapi/src/api/page/content-types/page/schema.json:1]: yes - dynamic section components are now alphabetically ordered when translated, this was one of the asks

File: strapi/src/components/sections/procurements.json (hunk 2 above)
Comment (jankaBehal) [reply strapi/src/components/sections/procurements.json:14]: This is intended as this section will include links as well, not only plain string text

File: strapi/src/components/sections/table.json
Comment (jankaBehal) [reply strapi/src/components/sections/table.json:1]: removed

File: strapi/src/components/sections/procurements.json (hunk 3 above)
Comment (jankaBehal) [reply strapi/src/components/sections/procurements.json]: done

File: next/src/services/josephine/fetchProcurements.ts (hunk 6 above)
Comment (jankaBehal) [reply next/src/services/josephine/fetchProcurements.ts]: done

File: next/src/pages/api/josephine.ts (hunk 8 above)
Comment (jankaBehal) [reply next/src/pages/api/josephine.ts]: done

File: next/src/pages/api/josephine.ts (hunk 9 above)
Comment (jankaBehal) [reply next/src/pages/api/josephine.ts]: done

File: next/src/components/sections/ProcurementsSection.tsx (hunk 11 above)
Comment (jankaBehal) [reply next/src/components/sections/ProcurementsSection.tsx]: you are right, fixed

**Comment by jankaBehal** [comment]
renaming of Procurements - Tenders will be in other PR

File: strapi/src/components/sections/procurements.json (hunk 2 above)
Comment (radoslavzeman) [reply strapi/src/components/sections/procurements.json:14]: Hmm okay I understand.

File: strapi/src/api/page/content-types/page/schema.json
Comment (radoslavzeman) [inline strapi/src/api/page/content-types/page/schema.json:1]: Please doublecheck if no section is missing. Ideally, reordering would be separate PR.

File: next/src/services/josephine/utils.tsx [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+        allColumns.includes(tenderEntry[0]),
+      ),
+    )
+
+    const detailLink = (
+      <Link
+        className="underline"
+        href={tender.tender_link}
+        target="_blank"
+        aria-label={detailString}
```
Comment (radoslavzeman) [inline next/src/services/josephine/utils.tsx:52]: Specifying the same text for aria-label and for link text is redundant.
If link value is a text, it's enough.
Please remove `aria-label`.

File: next/src/services/josephine/utils.tsx [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+  return procurements?.tenders.map((tender) => {
+    const tenderAttributes = Object.fromEntries(
+      Object.entries(tender).filter((tenderEntry: [key: string, value: string]) =>
+        allColumns.includes(tenderEntry[0]),
+      ),
+    )
+
+    const detailLink = (
+      <Link
+        className="underline"
```
Comment (radoslavzeman) [inline next/src/services/josephine/utils.tsx:49]: Please use our `Link` component and its relevant variant.

File: next/src/services/josephine/fetchProcurements.ts [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+    })
+
+    const tendersFiltered = tendersSorted.slice(
+      (currentPage - 1) * tendersPerPage,
+      currentPage * tendersPerPage,
+    )
+
+    obj = {
+      tenders: [...(tendersFiltered as ModifiedTender[])],
+      tendersCount: result?.tenders?.tender?.length,
```
[the remark this answers is not in the document]
Comment (jankaBehal) [reply next/src/services/josephine/fetchProcurements.ts:87]: this needs to be number of all possible options, not only displayed ones

File: strapi/src/api/page/content-types/page/schema.json
Comment (jankaBehal) [reply strapi/src/api/page/content-types/page/schema.json:1]: ko doublechecked with current master

File: next/.env.example (hunk 1 above)
Comment (jankaBehal) [reply next/.env.example]: I will solve it in next PR

**Review by radoslavzeman — COMMENTED** [review]
👍

**Review by jankaBehal — APPROVED** [review]
Good to be merged
